### DISCUSSION TITLE
[NBug] WaitForInputIdle failed.  This could be because the process...

### DISCUSSION THREAD

**Discussion by maxx27** [gitextensions/gitextensions#9840] [Q&A] [answered]
:warning: The sections below must be filled in and this text must be removed or the issue will be closed.

## Error Details
```
System.InvalidOperationException: WaitForInputIdle failed.  This could be because the process does not have a graphical interface.
   at System.Diagnostics.Process.WaitForInputIdle(Int32 milliseconds)
   at System.Diagnostics.Process.WaitForInputIdle()
   at GitCommands.Executable.ProcessWrapper.WaitForInputIdle()
   at GitCommands.GitModule.StartPageantWithKey(String sshKeyFile)
   at GitUI.BrowseForPrivateKey.LoadKey(IWin32Window parent, String path)
   at GitUI.BrowseForPrivateKey.BrowseAndLoad(IWin32Window parent)
   at GitUI.FormPuttyError.LoadSSHKey_Click(Object sender, EventArgs e)
   at System.Windows.Forms.Control.OnClick(EventArgs e)
   at System.Windows.Forms.Button.OnClick(EventArgs e)
   at System.Windows.Forms.Button.OnMouseUp(MouseEventArgs mevent)
   at System.Windows.Forms.Control.WmMouseUp(Message& m, MouseButtons button, Int32 clicks)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ButtonBase.WndProc(Message& m)
   at System.Windows.Forms.Button.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.OnMessage(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
```

## Additional information
Run git fetch. Window to load ssh key for putty appears. While I select the folder, I got the exception

## Environment
- Git Extensions 3.4.1.9675
- Build f49b4d059c5b135b4b2bf9eae4f443f19f6ac5df (Dirty)
- Git 2.27.0.windows.1
- Microsoft Windows NT 10.0.18363.0
- .NET Framework 4.8.4180.0
- DPI 96dpi (no scaling)

**Comment by RussKie** [comment]
Is this repros consistently? Steps to reproduce?

**Comment by mstv** [comment]
I guess there is a wrapper process inbetween, which has not started a message loop (yet).
Then `Process.WaitForInputIdle` times out while the subprocess asks for the ssh key.

Could you tell us the versions of PuTTY and Pageant as well as the configured PuTTY path, @maxx27?
Is your Pageant running in the background?
![grafik](https://user-images.githubusercontent.com/36601201/86495560-2a350900-bd7a-11ea-9398-41ed48186351.png)

**Comment by maxx27** [comment]
Hello Michael,

> Could you tell us the versions of PuTTY and Pageant as well as the configured PuTTY path

Here is the information from about dialog of the `pageant` and `putty` (it is same for both of them):
```
Release 0.74

Build platform: 64-bit x86 Windows
Compiler: clang 11.0.0 (https://github.com/llvm/llvm-project/ bc15bf66dcca76cc06fe71fca35b74dc4d521021), emulating Visual Studio 2013 (12.0), _MSC_VER=1800
Embedded HTML Help file: yes
Source commit: 014d4fb151369f255b3debed7d15a154fd9036f5
```

> Is your Pageant running in the background?

Yes. If it doesn't then GitExtentions runs it automatically.

A couple words about the environment:

* I have installed putty separately.
* My keyfile has a passphrase.
* In some cases I use [KiTTY](http://www.9bis.net/kitty/). I'm not sure that KiTTY was running when I got the error.

I tried different cases to reproduce the error but I can't. I know that I faced it a couple of times. When I will face the error again then I will provide more information.

**Comment by MartinWilkerson** [comment, marked as answer]
I would like to add to this as I think I'm getting the same error. To reproduce I have to:

1. Install PuTTY (and pageant) through Chocolatey, which creates shims for the actual executables using shimgen (https://docs.chocolatey.org/en-us/features/shim)
2. Make sure that pageant is not running at all (no process containing PAGEANT running)
3. Have a remote repository configured with an ssh key configured for auto-loading
4. Fetch, pull, push or otherwise perform an operation on the remote repository using git extensions

At this point the error message described above occurs.

Step 2 is required because the place the error is thrown is in `GitModule.cs` line 1447:
```csharp
            if (!IsPageantRunning())
            {
                // NOTE we leave the process to dangle here
                var process = pageantExecutable.Start("");

                process.WaitForInputIdle();
            }
```

I have fixed it by updating my paths in git extensions' config to point to the actual binaries, not the shims generated by Chocolatey. To find them I ran `pageant --shimgen-log` which outputs the actual location the shim points to.

**Reply by ronaldkroon** [reply]
This fix worked for me as well. Thanks Martin!
Would it be possible to update Git Extensions so that this manual fix would not be required?
